backends/bmv2/common/action.cpp

```diff
@@ -47,6 +47,36 @@ void ActionConverter::convertActionBody(const IR::Vector<IR::StatOrDecl>* body,
         // TODO(jafingerhut) - add line/col at all individual cases below,
         // or perhaps it can be done as a common case above or below
         // for all of them?
+
+        IR::MethodCallExpression *mce2;
+        auto isR = false;
+        if (s->is<IR::AssignmentStatement>()) {
+            auto assign = s->to<IR::AssignmentStatement>();
+            const IR::Expression *l, *r;
+            l = assign->left;
+            r = assign->right;
+            auto mce = r->to<IR::MethodCallExpression>();
+            if (mce != nullptr) {
+                auto mi = P4::MethodInstance::resolve(mce, ctxt->refMap, ctxt->typeMap);
+                auto em = mi->to<P4::ExternMethod>();
+                if (em != nullptr) {
```

Comment:
Does some pass make sure that reads are only used in assignments on the rhs?
You can have an expression `a = 2 + r.read(5)`

Reply:
Thanks! I am adding a pass to psa backend to convert all `r.read()` into  `dst = r.read(index)` format. So that expression of arbitrary complexity can be handled.

Reply:
Actually, I just realized one of the frontend passes `SideEffectOrdering` ensures all r.read will be transformed into the format `dst = r.read(index)`
e.g. `a = r.read(1) + r.read(2) - 5` will be translated as  
`tmp_0 = r.read(1);`  
`tmp_1 = r.read(2);`  
`a = tmp_0 + tmp_1 - 5;`